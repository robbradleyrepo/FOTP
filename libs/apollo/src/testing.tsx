import { MockedResponse, MockLink } from "@apollo/react-testing";
import { ApolloClientType, NextPageWithApollo } from "@sss/apollo";
import { Locale } from "@sss/i18n";
import { getMockedRouter } from "@sss/next/testing";
import {
  act,
  render,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import {
  InMemoryCache,
  InMemoryCacheConfig,
  IntrospectionFragmentMatcher,
  IntrospectionResultData,
} from "apollo-cache-inmemory";
import { ApolloClient } from "apollo-client";
import {
  GetServerSideProps,
  GetServerSidePropsContext,
  GetStaticProps,
} from "next";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { Router } from "next/router";
import { ParsedUrlQuery, stringify } from "querystring";
import React, { ComponentType } from "react";

import { initializeApollo } from "./client";
jest.mock("./client");

interface BaseSandbox {
  // The global mocks to apply
  globalMocks?: MockedResponse[];
  // Additional mocks under test
  mocks?: MockedResponse[];
  renderOptions?: Omit<RenderOptions, "queries">;
  pathname?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params?: { [key: string]: any };
}

interface PageSandbox extends BaseSandbox {
  NextPage: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    default: React.ComponentType<any>;
    getStaticProps?: GetStaticProps;
    getServerSideProps?: GetServerSideProps;
  };
}

interface PageResult {
  mockLink: MockLink;
  redirect?: { permanent: boolean; destination: string };
  notFound?: boolean;
  page: RenderResult;
  router: Router;
}

interface ComponentSandbox extends BaseSandbox {
  Component: React.ReactNode;
}

interface ComponentResult {
  mockLink: MockLink;
  component: RenderResult;
  router: Router;
}

interface SandboxInit {
  globalMocks?: MockedResponse[];
  introspectionResultData?: IntrospectionResultData;
  locale: Locale;
  Wrapper: ComponentType<{
    apolloClient: ApolloClientType;
    locale: Locale;
  }>;
}

export const createApolloClient = (
  mocks: MockedResponse[],
  introspectionQueryResultData?: IntrospectionResultData
) => {
  const cacheConfig: InMemoryCacheConfig = { addTypename: true };

  // `InMemoryCache` doesn't like `fragmentMatcher: undefined`. We must omit it.
  if (introspectionQueryResultData) {
    cacheConfig.fragmentMatcher = new IntrospectionFragmentMatcher({
      introspectionQueryResultData,
    });
  }

  return new ApolloClient({
    cache: new InMemoryCache(cacheConfig),
    link: new MockLink(mocks, true),
  });
};

export default class Sandbox {
  public constructor(private opts: SandboxInit) {}

  protected createApolloClient(
    mocks: MockedResponse[] = [],
    globalMocks: MockedResponse[] = this.opts.globalMocks ?? []
  ) {
    return createApolloClient(
      [...globalMocks, ...mocks],
      this.opts.introspectionResultData
    );
  }

  public async page({
    NextPage: NextPageBundle,
    globalMocks,
    mocks,
    renderOptions,
    pathname = "/",
    params = {},
  }: PageSandbox): Promise<PageResult> {
    const router = getMockedRouter(pathname, params);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).window = Object.create(window);
    Object.defineProperty(window, "location", {
      value: {
        href: `https://localhost${router.asPath}`,
        pathname: router.asPath,
        search: params ? stringify(params) : "",
      },
      writable: true,
    });

    const apolloClient = this.createApolloClient(mocks, globalMocks);

    (initializeApollo as jest.Mock).mockReturnValueOnce(apolloClient);

    const { default: PageComponent } = NextPageBundle;

    const pageData = NextPageBundle.getStaticProps
      ? await NextPageBundle.getStaticProps({ locale: "en-US", params })
      : NextPageBundle.getServerSideProps
      ? await NextPageBundle.getServerSideProps(
          // `getServerSideProps` context argument also requires
          // `req: IncomingMessage` and `res: ServerResponse`. Cast
          // this for testing purposes.
          ({
            locale: "en-US",
            params,
            req: {},
            res: {
              setHeader: jest.fn(),
              statusCode: null,
            },
          } as unknown) as GetServerSidePropsContext<ParsedUrlQuery>
        )
      : { props: {} };

    let page;

    act(() => {
      const Wrapper = this.opts.Wrapper;
      page = render(
        "props" in pageData &&
          !("notFound" in pageData) &&
          !("redirect" in pageData) ? (
          <RouterContext.Provider value={router}>
            <Wrapper apolloClient={apolloClient} locale={this.opts.locale}>
              <PageComponent {...pageData.props} />
            </Wrapper>
          </RouterContext.Provider>
        ) : (
          <div />
        ),
        {
          ...renderOptions,
        }
      );
    });

    return {
      link: apolloClient.link,
      notFound: "notFound" in pageData && pageData.notFound,
      page,
      redirect: "redirect" in pageData && pageData.redirect,
      router,
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
  }

  public async component({
    Component,
    ...options
  }: ComponentSandbox): Promise<ComponentResult> {
    const NextPageComponent: NextPageWithApollo = () => {
      return <>{Component}</>;
    };

    const { page: component, ...rest } = await this.page({
      NextPage: { default: NextPageComponent },
      ...options,
    });

    return {
      component,
      ...rest,
    };
  }
}
