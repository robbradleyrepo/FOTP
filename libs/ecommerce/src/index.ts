export { default as ECOMMERCE } from "./config";
export type { CountryCodes } from "./countries";
export { default as Countries } from "./countries";
export { ProductSchema, useProductSchemaJson } from "./jsonschema";
export { getState, getStateCode, default as USStates } from "./us-states";
