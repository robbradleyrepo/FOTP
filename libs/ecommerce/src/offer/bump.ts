import { useQuery } from "@apollo/react-hooks";
import { captureException } from "@sentry/nextjs";
import { useRouter } from "next/router";
import { ParsedUrlQuery } from "querystring";
import { useMemo } from "react";

import {
  CartData,
  CartStatus,
  hasCompleteLineItemData,
  useCart,
} from "../cart";
import {
  findVariantBySku,
  Product,
  PRODUCT_BY_HANDLE,
  ProductData,
  Variant,
} from "../product";

export interface BumpOffer {
  exclude?: { handles: string[] }; // Don't trigger offer if the cart contains a matching product
  include: { handles: string[] };
  product: Record<"handle" | "sku", string>;
}

interface GetApplicableBumpOfferParams {
  cart: Pick<CartData, "lineItems" | "status">;
  offers: BumpOffer[];
  product: Pick<Product, "handle">;
  query: ParsedUrlQuery;
}

export const getApplicableBumpOffer = ({
  cart: { lineItems, status },
  offers,
  product,
  query,
}: GetApplicableBumpOfferParams): BumpOffer | null => {
  let applicableOffer = null;

  if (
    offers &&
    query.bump_offers !== "false" &&
    hasCompleteLineItemData(lineItems, status) // We can't tell if the offer is applicable or not until all cart data is available
  ) {
    const isApplicableOffer = (offer: BumpOffer) =>
      !lineItems.some(
        ({ variant }) =>
          offer.product.handle === variant.product.handle || // The offer is not applicable if the cart contains the offer product
          (offer.exclude?.handles ?? []).includes(variant.product.handle) // Or if the cart contains an explicitly excluded product
      );

    // Loop through all offers and find the first that applies
    for (const offer of offers) {
      if (
        offer.include.handles.includes(product.handle) &&
        isApplicableOffer(offer)
      ) {
        applicableOffer = offer;
        break;
      }
    }
  }

  return applicableOffer;
};

export interface UseBumpOfferOptions {
  disabled?: boolean;
  ignoreCart?: boolean;
  offers: BumpOffer[];
  product: Pick<Product, "handle">;
}

export enum BumpOfferStatus {
  ERROR = "ERROR",
  FETCHING = "FETCHING",
  INITIALIZING = "INITIALIZING",
  READY = "READY",
}

export type UseBumpOfferResult =
  | {
      offer: BumpOffer | null;
      status: BumpOfferStatus;
    }
  | {
      offer: BumpOffer;
      product: Product;
      status: BumpOfferStatus.READY;
      variant: Variant;
    };

export const useBumpOffer = ({
  disabled = false,
  ignoreCart,
  offers,
  product,
}: UseBumpOfferOptions): UseBumpOfferResult => {
  const { lineItems, status } = useCart();
  const router = useRouter();

  const effectiveLineItems = ignoreCart ? null : lineItems;

  const offer = useMemo(
    () =>
      disabled
        ? null
        : getApplicableBumpOffer({
            cart: { lineItems: effectiveLineItems ?? [], status },
            offers,
            product,
            query: router.query,
          }),
    [disabled, effectiveLineItems, offers, product, router, status]
  );

  const queryResult = useQuery<ProductData>(PRODUCT_BY_HANDLE, {
    skip: !offer?.product.handle,
    variables: { handle: offer?.product.handle },
  });

  return useMemo(() => {
    if (status !== CartStatus.READY) {
      return {
        offer: null,
        status: BumpOfferStatus.INITIALIZING,
      };
    }

    const { data, loading } = queryResult;

    if (!offer) {
      return {
        offer: null,
        status: BumpOfferStatus.READY,
      };
    }

    if (data?.product && !loading) {
      const { product } = data;
      const variant = findVariantBySku(product.variants, offer.product.sku);

      if (variant) {
        if (!variant.availableForSale) {
          return {
            offer: null,
            status: BumpOfferStatus.READY,
          };
        }

        return {
          offer,
          product,
          status: BumpOfferStatus.READY,
          variant,
        };
      }

      const error = new Error("Invalid bump offer configuration");

      if (process.env.NODE_ENV === "production") {
        captureException(error, { extra: { offer, product } });
      } else {
        throw error;
      }
    }

    if (loading) {
      return {
        offer,
        status: BumpOfferStatus.FETCHING,
      };
    }

    return {
      offer,
      status: BumpOfferStatus.ERROR,
    };
  }, [offer, queryResult, status]);
};
