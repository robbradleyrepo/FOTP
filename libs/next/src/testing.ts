import { Router } from "next/router";
import { ParsedUrlQuery } from "querystring";
import { mocked } from "ts-jest/utils";

import { getAsPath } from "./helpers";

export const getMockedRouter = (pathname: string, params?: ParsedUrlQuery) => {
  const asPath = getAsPath(pathname, params);

  return mocked(({
    asPath,
    events: {
      off: jest.fn(),
      on: jest.fn(),
    },
    isReady: true,
    pathname,
    prefetch: jest.fn(async () => undefined),
    push: jest.fn(),
    query: params,
    replace: jest.fn(),
    route: pathname,
  } as unknown) as Router);
};
