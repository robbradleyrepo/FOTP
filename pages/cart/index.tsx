export { default, getStaticProps } from "../../src/pages/cart";
