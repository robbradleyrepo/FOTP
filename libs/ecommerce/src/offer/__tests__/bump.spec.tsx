import {
  MockedProvider as MockApolloProvider,
  MockedResponse,
} from "@apollo/react-testing";
import * as ecommerce from "@sss/ecommerce/testing";
import { renderHook } from "@testing-library/react-hooks";
import React, { FC } from "react";

import { CartProvider, CartStatus } from "../../cart";
import { BumpOffer, getApplicableBumpOffer, useBumpOffer } from "../bump";

jest.mock("next/router", () => ({
  useRouter: jest.fn(() => ({ query: {} })),
}));

describe("getApplicableBumpOffer", () => {
  const bar: BumpOffer = {
    include: {
      handles: ["baz", "foo"],
    },
    product: { handle: "bar", sku: "bar-x1" },
  };
  const baz: BumpOffer = {
    include: {
      handles: ["bar", "foo"],
    },
    product: { handle: "baz", sku: "baz-x1" },
  };
  const foo: BumpOffer = {
    exclude: { handles: ["fooish"] },
    include: { handles: ["bar", "baz"] },
    product: { handle: "foo", sku: "foo-x1" },
  };

  const offers = [foo, bar, baz];

  it("should return the first offer that applies to the provided product if there are no items in the cart", () => {
    expect(
      getApplicableBumpOffer({
        cart: { lineItems: [], status: CartStatus.READY },
        offers,
        product: { handle: "foo" },
        query: {},
      })
    ).toBe(bar);
  });

  it("should not return an offer for a product that is already in the cart", () => {
    expect(
      getApplicableBumpOffer({
        cart: {
          lineItems: [{ variant: { product: { handle: "bar" } } }],
          status: CartStatus.READY,
        } as any, // eslint-disable-line @typescript-eslint/no-explicit-any
        offers,
        product: { handle: "foo" },
        query: {},
      })
    ).toBe(baz);

    expect(
      getApplicableBumpOffer({
        cart: {
          lineItems: [
            { variant: { product: { handle: "bar" } } },
            { variant: { product: { handle: "baz" } } },
          ],
          status: CartStatus.READY,
        } as any, // eslint-disable-line @typescript-eslint/no-explicit-any
        offers,
        product: { handle: "foo" },
        query: {},
      })
    ).toBeNull();
  });

  it("should not return an offer if the cart contains a blacklisted product", () => {
    const params = {
      cart: {
        lineItems: [{ variant: { product: { handle: "blah" } } }],
        status: CartStatus.READY,
      } as any, // eslint-disable-line @typescript-eslint/no-explicit-any
      offers,
      product: { handle: "bar" },
      query: {},
    };

    expect(getApplicableBumpOffer(params)).toBe(foo);

    expect(
      getApplicableBumpOffer({
        ...params,
        cart: {
          ...params.cart,
          lineItems: [{ variant: { product: { handle: "fooish" } } }] as any[], // eslint-disable-line @typescript-eslint/no-explicit-any
        },
      })
    ).toBe(baz);
  });

  it("should not return any offers if complete cart data is not available", () => {
    expect(
      getApplicableBumpOffer({
        cart: { lineItems: [], status: CartStatus.FETCHING },
        offers,
        product: { handle: "foo" },
        query: {},
      })
    ).toBeNull();
  });
});

describe("useBumpOffer", () => {
  const NODE_ENV = process.env.NODE_ENV;

  afterEach(() => {
    (process.env.NODE_ENV as unknown) = NODE_ENV;
  });

  const mocks: MockedResponse[] = [
    {
      delay: 1,
      ...ecommerce.mockedResponses.product.productByHandle,
    },
  ];

  const wrapper: FC = ({ children }) => (
    <MockApolloProvider
      mocks={mocks}
      addTypename={false}
      defaultOptions={{
        query: { fetchPolicy: "no-cache" },
        watchQuery: { fetchPolicy: "no-cache" },
      }}
    >
      <CartProvider>{children}</CartProvider>
    </MockApolloProvider>
  );

  it("should return the first applicable bump offer and fetch its product data", async () => {
    const { result, waitForNextUpdate } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "move", sku: "move" },
            },
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "fake", sku: "fake" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    await waitForNextUpdate();

    expect(result.all).toMatchSnapshot();
  });

  it("should return null for the bump offer if the applicable offer is unavailable for sale", async () => {
    const { result, waitForNextUpdate } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "move", sku: "move" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    await waitForNextUpdate();

    expect(result.all).toMatchSnapshot();
  });

  it("should return an error status if the data fails to load", async () => {
    const { result, waitForNextUpdate } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "missing", sku: "missing" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    await waitForNextUpdate();

    expect(result.all).toMatchSnapshot();
  });

  it("should return an error status in production if the bump product does not include the specified variant", async () => {
    (process.env.NODE_ENV as string) = "production";

    const { result, waitForNextUpdate } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "move", sku: "missing" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    await waitForNextUpdate();

    expect(result.all).toMatchSnapshot();
  });

  it("should return throw an error in a non-production environment if the bump product does not include the specified variant", async () => {
    const { result, waitForNextUpdate } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["foo"],
              },
              product: { handle: "move", sku: "missing" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    await waitForNextUpdate();

    expect(() => result.current).toThrow();
  });

  it("should return a `null` offer if there is no valid offer for the current product", async () => {
    const { result } = renderHook(
      () =>
        useBumpOffer({
          offers: [
            {
              include: {
                handles: ["bar"],
              },
              product: { handle: "move", sku: "move" },
            },
          ],
          product: { handle: "foo" },
        }),
      { wrapper }
    );

    expect(result.all).toMatchSnapshot();
  });
});
