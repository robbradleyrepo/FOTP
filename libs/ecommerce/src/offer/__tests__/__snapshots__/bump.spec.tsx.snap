// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBumpOffer should return a \`null\` offer if there is no valid offer for the current product 1`] = `
Array [
  Object {
    "offer": null,
    "status": "INITIALIZING",
  },
  Object {
    "offer": null,
    "status": "READY",
  },
]
`;

exports[`useBumpOffer should return an error status if the data fails to load 1`] = `
Array [
  Object {
    "offer": null,
    "status": "INITIALIZING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "missing",
        "sku": "missing",
      },
    },
    "status": "FETCHING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "missing",
        "sku": "missing",
      },
    },
    "status": "ERROR",
  },
]
`;

exports[`useBumpOffer should return an error status in production if the bump product does not include the specified variant 1`] = `
Array [
  Object {
    "offer": null,
    "status": "INITIALIZING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "missing",
      },
    },
    "status": "FETCHING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "missing",
      },
    },
    "status": "ERROR",
  },
]
`;

exports[`useBumpOffer should return null for the bump offer if the applicable offer is unavailable for sale 1`] = `
Array [
  Object {
    "offer": null,
    "status": "INITIALIZING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "move",
      },
    },
    "status": "FETCHING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "move",
      },
    },
    "product": Object {
      "__typename": "Product",
      "availableForSale": true,
      "bottomline": Object {
        "__typename": "ReviewBottomline",
        "averageScore": 4.5,
        "totalReviews": 2,
      },
      "bundleUnit": null,
      "bundleUnitPlural": null,
      "caloriesPerUnit": null,
      "compareAtPriceV2": null,
      "containers": null,
      "defaultSelectionSku": null,
      "defaultSelectionSubscription": null,
      "defaultVariantOrder": null,
      "description": "Targeted joint and mobility support",
      "featuredImage": Object {
        "__typename": "ImageConnection",
        "edges": Array [
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "id": "1",
              "src": "https://cdn.shopify.com/mock-image/product-1.jpg",
            },
          },
        ],
      },
      "handle": "move",
      "hasSubscription": null,
      "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
      "images": Object {
        "__typename": "ImageConnection",
        "edges": Array [
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "1",
              "url": "https://cdn.shopify.com/mock-image/product-1.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "2",
              "url": "https://cdn.shopify.com/mock-image/product-2.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "3",
              "url": "https://cdn.shopify.com/mock-image/product-3.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "4",
              "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
              "width": 1000,
            },
          },
        ],
      },
      "isSubscriptionOnly": null,
      "listingSku": null,
      "listingSubscription": null,
      "listingSubtitle": null,
      "options": Array [
        Object {
          "__typename": "ProductOption",
          "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0T3B0aW9uLzU4NzMxNjM4NjIwOTQ=",
          "name": "Title",
          "values": Array [
            "Default Title",
          ],
        },
      ],
      "preorderShippingEstimate": null,
      "preorderType": null,
      "priceRange": Object {
        "__typename": "ProductPriceRange",
        "maxVariantPrice": Object {
          "__typename": "MoneyV2",
          "amount": "95.98",
          "currencyCode": "USD",
        },
        "minVariantPrice": Object {
          "__typename": "MoneyV2",
          "amount": "35.99",
          "currencyCode": "USD",
        },
      },
      "productType": "Pet Vitamins & Supplements",
      "seo": Object {
        "description": "",
        "title": "",
      },
      "subscriptionFrequencies": null,
      "subscriptionUnit": null,
      "subtitle": Object {
        "__typename": "Metafield",
        "id": "metafield-subtitle-move",
        "value": "Targeted joint and mobility support",
      },
      "title": "Move",
      "totalInventory": 1000,
      "unit": null,
      "variants": Object {
        "__typename": "ProductVariantConnection",
        "edges": Array [
          Object {
            "__typename": "ProductVariantEdge",
            "node": Object {
              "__typename": "ProductVariant",
              "availableForSale": true,
              "barcode": "123",
              "bundleSize": null,
              "compareAtPriceV2": Object {
                "__typename": "MoneyV2",
                "amount": "39.99",
                "currencyCode": "USD",
              },
              "defaultShippingIntervalFrequency": Object {
                "__typename": "Metafield",
                "id": "defaultShippingIntervalFrequency-1",
                "value": "1",
              },
              "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
              "image": Object {
                "__typename": "Image",
                "height": 1000,
                "id": "4",
                "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                "width": 1000,
              },
              "listingUsp": null,
              "priceV2": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "quantityAvailable": 10,
              "selectedOptions": Array [
                Object {
                  "__typename": "SelectedOption",
                  "name": "Title",
                  "value": "Default Title",
                },
              ],
              "sku": "move",
              "subscriptionPriceAmount": Object {
                "__typename": "Metafield",
                "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                "value": "29.99",
              },
              "title": "Default Title",
              "units": Object {
                "__typename": "Metafield",
                "id": "gid://shopify/Metafield/1",
                "value": "60",
              },
            },
          },
        ],
      },
      "vendor": "Front Of The Pack",
    },
    "status": "READY",
    "variant": Object {
      "__typename": "ProductVariant",
      "availableForSale": true,
      "barcode": "123",
      "bundleSize": null,
      "compareAtPriceV2": Object {
        "__typename": "MoneyV2",
        "amount": "39.99",
        "currencyCode": "USD",
      },
      "defaultShippingIntervalFrequency": Object {
        "__typename": "Metafield",
        "id": "defaultShippingIntervalFrequency-1",
        "value": "1",
      },
      "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
      "image": Object {
        "__typename": "Image",
        "height": 1000,
        "id": "4",
        "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
        "width": 1000,
      },
      "listingUsp": null,
      "priceV2": Object {
        "__typename": "MoneyV2",
        "amount": "34.99",
        "currencyCode": "USD",
      },
      "quantityAvailable": 10,
      "selectedOptions": Array [
        Object {
          "__typename": "SelectedOption",
          "name": "Title",
          "value": "Default Title",
        },
      ],
      "sku": "move",
      "subscriptionPriceAmount": Object {
        "__typename": "Metafield",
        "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
        "value": "29.99",
      },
      "title": "Default Title",
      "units": Object {
        "__typename": "Metafield",
        "id": "gid://shopify/Metafield/1",
        "value": "60",
      },
    },
  },
]
`;

exports[`useBumpOffer should return the first applicable bump offer and fetch its product data 1`] = `
Array [
  Object {
    "offer": null,
    "status": "INITIALIZING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "move",
      },
    },
    "status": "FETCHING",
  },
  Object {
    "offer": Object {
      "include": Object {
        "handles": Array [
          "foo",
        ],
      },
      "product": Object {
        "handle": "move",
        "sku": "move",
      },
    },
    "product": Object {
      "__typename": "Product",
      "availableForSale": true,
      "bottomline": Object {
        "__typename": "ReviewBottomline",
        "averageScore": 4.5,
        "totalReviews": 2,
      },
      "bundleUnit": null,
      "bundleUnitPlural": null,
      "caloriesPerUnit": null,
      "compareAtPriceV2": null,
      "containers": null,
      "defaultSelectionSku": null,
      "defaultSelectionSubscription": null,
      "defaultVariantOrder": null,
      "description": "Targeted joint and mobility support",
      "featuredImage": Object {
        "__typename": "ImageConnection",
        "edges": Array [
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "id": "1",
              "src": "https://cdn.shopify.com/mock-image/product-1.jpg",
            },
          },
        ],
      },
      "handle": "move",
      "hasSubscription": null,
      "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
      "images": Object {
        "__typename": "ImageConnection",
        "edges": Array [
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "1",
              "url": "https://cdn.shopify.com/mock-image/product-1.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "2",
              "url": "https://cdn.shopify.com/mock-image/product-2.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "3",
              "url": "https://cdn.shopify.com/mock-image/product-3.jpg",
              "width": 1000,
            },
          },
          Object {
            "__typename": "ImageEdge",
            "node": Object {
              "__typename": "Image",
              "height": 1000,
              "id": "4",
              "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
              "width": 1000,
            },
          },
        ],
      },
      "isSubscriptionOnly": null,
      "listingSku": null,
      "listingSubscription": null,
      "listingSubtitle": null,
      "options": Array [
        Object {
          "__typename": "ProductOption",
          "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0T3B0aW9uLzU4NzMxNjM4NjIwOTQ=",
          "name": "Title",
          "values": Array [
            "Default Title",
          ],
        },
      ],
      "preorderShippingEstimate": null,
      "preorderType": null,
      "priceRange": Object {
        "__typename": "ProductPriceRange",
        "maxVariantPrice": Object {
          "__typename": "MoneyV2",
          "amount": "95.98",
          "currencyCode": "USD",
        },
        "minVariantPrice": Object {
          "__typename": "MoneyV2",
          "amount": "35.99",
          "currencyCode": "USD",
        },
      },
      "productType": "Pet Vitamins & Supplements",
      "seo": Object {
        "description": "",
        "title": "",
      },
      "subscriptionFrequencies": null,
      "subscriptionUnit": null,
      "subtitle": Object {
        "__typename": "Metafield",
        "id": "metafield-subtitle-move",
        "value": "Targeted joint and mobility support",
      },
      "title": "Move",
      "totalInventory": 1000,
      "unit": null,
      "variants": Object {
        "__typename": "ProductVariantConnection",
        "edges": Array [
          Object {
            "__typename": "ProductVariantEdge",
            "node": Object {
              "__typename": "ProductVariant",
              "availableForSale": true,
              "barcode": "123",
              "bundleSize": null,
              "compareAtPriceV2": Object {
                "__typename": "MoneyV2",
                "amount": "39.99",
                "currencyCode": "USD",
              },
              "defaultShippingIntervalFrequency": Object {
                "__typename": "Metafield",
                "id": "defaultShippingIntervalFrequency-1",
                "value": "1",
              },
              "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
              "image": Object {
                "__typename": "Image",
                "height": 1000,
                "id": "4",
                "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                "width": 1000,
              },
              "listingUsp": null,
              "priceV2": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "quantityAvailable": 10,
              "selectedOptions": Array [
                Object {
                  "__typename": "SelectedOption",
                  "name": "Title",
                  "value": "Default Title",
                },
              ],
              "sku": "move",
              "subscriptionPriceAmount": Object {
                "__typename": "Metafield",
                "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                "value": "29.99",
              },
              "title": "Default Title",
              "units": Object {
                "__typename": "Metafield",
                "id": "gid://shopify/Metafield/1",
                "value": "60",
              },
            },
          },
        ],
      },
      "vendor": "Front Of The Pack",
    },
    "status": "READY",
    "variant": Object {
      "__typename": "ProductVariant",
      "availableForSale": true,
      "barcode": "123",
      "bundleSize": null,
      "compareAtPriceV2": Object {
        "__typename": "MoneyV2",
        "amount": "39.99",
        "currencyCode": "USD",
      },
      "defaultShippingIntervalFrequency": Object {
        "__typename": "Metafield",
        "id": "defaultShippingIntervalFrequency-1",
        "value": "1",
      },
      "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
      "image": Object {
        "__typename": "Image",
        "height": 1000,
        "id": "4",
        "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
        "width": 1000,
      },
      "listingUsp": null,
      "priceV2": Object {
        "__typename": "MoneyV2",
        "amount": "34.99",
        "currencyCode": "USD",
      },
      "quantityAvailable": 10,
      "selectedOptions": Array [
        Object {
          "__typename": "SelectedOption",
          "name": "Title",
          "value": "Default Title",
        },
      ],
      "sku": "move",
      "subscriptionPriceAmount": Object {
        "__typename": "Metafield",
        "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
        "value": "29.99",
      },
      "title": "Default Title",
      "units": Object {
        "__typename": "Metafield",
        "id": "gid://shopify/Metafield/1",
        "value": "60",
      },
    },
  },
]
`;
